util/math_reference.h

```diff
@@ -15,22 +15,178 @@
 
 namespace reference {
 /* two argument relational reference */
-int isequal(float x, float y);
-int isnotequal(float x, float y);
-int isgreater(float x, float y);
-int isgreaterequal(float x, float y);
-int isless(float x, float y);
-int islessequal(float x, float y);
-int islessgreater(float x, float y);
-int isordered(float x, float y);
-int isunordered(float x, float y);
+template <typename T>
+auto isequal(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<std::equal_to>(a, b);
+}
+
+template <typename T>
+auto isnotequal(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<std::not_equal_to>(a, b);
+}
+
+template <typename T>
+auto isgreater(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<std::greater>(a, b);
+}
+
+template <typename T>
+auto isgreaterequal(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<std::greater_equal>(a, b);
+}
+
+template <typename T>
+auto isless(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<std::less>(a, b);
+}
+
+template <typename T>
+auto islessequal(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<std::less_equal>(a, b);
+}
+
+template<typename T>
+struct islessgreater_func{
+  constexpr bool operator()(const T& x, const T& y) const {
+    return (x < y) || (x > y);
+  }
+};
+template <typename T>
+auto islessgreater(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<islessgreater_func>(a, b);
+}
+
+template<typename T>
+struct isordered_func{
+  constexpr bool operator()(const T& x, const T& y) const {
+   return (x == x) && (y == y);
+  }
+};
+template <typename T>
+auto isordered(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<isordered_func>(a, b);
+}
+
+template<typename T>
+struct isunordered_func{
+  constexpr bool operator()(const T& x, const T& y) const {
+   return !((x == x) && (y == y));
+  }
+};
+template <typename T>
+auto isunordered(T a, T b) {
+  return sycl_cts::math::rel_func_dispatcher<isunordered_func>(a, b);
+}
 
 /* one argument relational reference */
-int isfinite(float x);
-int isinf(float x);
-int isnan(float x);
-int isnormal(float x);
-int signbit(float x);
+template<typename T>
+struct isfinite_func{
+  constexpr bool operator()(const T& x) const {
+   return std::isfinite(x);
+  }
+};
```

Comment:
Not fundamental, but you could be lazy for all these functions everywhere with C++14:
```suggestion
auto constexpr isfinite_func = [] (const auto& x) { return std::isfinite(x); };
```

Reply:
Changed to lambda and added dispatcher to work with it.